snapcraft/cli/store.py

```diff
@@ -252,18 +278,91 @@ def list_registered():
+@click.argument('login_file', metavar='FILE', type=click.File('w'))
+@click.option('--packages', metavar='<packages>',
+              help='Comma-separated list of packages to limit token access')
+@click.option('--channels', metavar='<channels>',
+              help='Comma-separated list of channels to limit token access')
+@click.option('--acls', metavar='<acls>',
+              help='Comma-separated list of ACLs to limit token access. Valid '
+                   "ACLs are 'package_upload', 'package_access', and "
+                   "'package_manage'")
+def export_login(login_file: TextIO, packages: str, channels: str, acls: str):
+    """Save attenuated login configuration for a store account in FILE.
+
+    This file can then be used to login to the given account with the
+    permissions specified.
+
+    For example, to limit access to the edge channel of any snap to which the
+    account has access:
+
+        snapcraft export-login --channels=edge
+
+    Or to limit access to only the edge channel of a single snap:
+
+        snapcraft export-login --packages=my-snap --channels=edge
+    """
+
+    package_list = None
+    channel_list = None
+    acl_list = None
+
+    if packages:
+        package_list = []
+        for package in packages.split(','):
+            package_list.append({'name': package, 'series': '16'})
+
+    if channels:
+        channel_list = channels.split(',')
+
+    if acls:
+        acl_list = acls.split(',')
+
+    store = storeapi.StoreClient()
+    if not snapcraft.login(store=store,
+                           packages=package_list,
+                           channels=channel_list,
+                           acls=acl_list,
+                           save=False):
+        sys.exit(1)
+
+    store.conf.save(config_fd=login_file)
+
+    print()
+    echo.info(
+        'Login successfully exported to {0!r}. This file can now be used with '
+        "'snapcraft login --with {0}' to login to this account with no "
+        'password and have the following capabilities:\n'.format(
+            login_file.name))
+    echo.info(_human_readable_acls(store))
+    echo.warning(
+        'This exported login is not encrypted. Do not commit it to version '
+        'control!')
+
+
 @storecli.command()
-def login():
+@click.option('--with', 'login_file', metavar='<login file>',
+              type=click.File('r'),
+              help="Path to file created with 'snapcraft export-login'")
+def login(login_file):
     """Login with your Ubuntu One e-mail address and password.
 
     If you do not have an Ubuntu One account, you can create one at
     https://dashboard.snapcraft.io/openid/login
     """
-    if not snapcraft.login():
+    store = storeapi.StoreClient()
+    if not snapcraft.login(store=store, config_fd=login_file):
         sys.exit(1)
 
     print()
-    echo.info('Login successful.')
+
+    if login_file:
+        echo.info(
+            'Login successful. You now have the following capabilities:\n')
```

Comment:
As above, “the following” would be shorter as “these”.

Reply:
Fixed.